jolt-core/src/zkvm/spartan/shift.rs

```diff
@@ -398,3 +310,455 @@ impl<F: JoltField, T: Transcript> SumcheckInstance<F, T> for ShiftSumcheck<F> {
+                for i in 0..p.len() / 2 {
+                    let p_evals = p.sumcheck_evals_array::<DEGREE>(i, BindingOrder::LowToHigh);
+                    let q_evals = q.sumcheck_evals_array::<DEGREE>(i, BindingOrder::LowToHigh);
+                    evals = array::from_fn(|i| evals[i] + p_evals[i] * q_evals[i]);
+                }
+                evals
+            })
+            .reduce(
+                || [F::zero(); DEGREE],
+                |a, b| array::from_fn(|i| a[i] + b[i]),
+            )
+            .to_vec()
+    }
+
+    fn bind(mut self, r_j: F::Challenge) -> Prover<F> {
+        self.sumcheck_challenges.push(r_j);
+
+        // Transition to phase 2.
+        if self.prefix_suffix_pairs[0].0.len().ilog2() == 1 {
+            return Prover::Phase2(Phase2Prover::gen(
+                &self.trace,
+                &self.bytecode_preprocessing,
+                &self.sumcheck_challenges,
+                self.gamma_powers,
+                &self.r_cycle,
+                &self.r_product,
+            ));
+        }
+
+        self.prefix_suffix_pairs.iter_mut().for_each(|(p, q)| {
+            p.bind(r_j, BindingOrder::LowToHigh);
+            q.bind(r_j, BindingOrder::LowToHigh);
+        });
+        Prover::Phase1(self)
+    }
+}
+
+/// Prover for 2nd half of the rounds.
+#[derive(Default, Allocative)]
+struct Phase2Prover<F: JoltField> {
+    unexpanded_pc_poly: MultilinearPolynomial<F>,
+    pc_poly: MultilinearPolynomial<F>,
+    is_virtual_poly: MultilinearPolynomial<F>,
+    is_first_in_sequence_poly: MultilinearPolynomial<F>,
+    is_noop_poly: MultilinearPolynomial<F>,
+    eq_plus_one_r_cycle: MultilinearPolynomial<F>,
+    eq_plus_one_r_product: MultilinearPolynomial<F>,
+    gamma_powers: [F; 5],
+}
+
+impl<F: JoltField> Phase2Prover<F> {
+    fn gen(
+        trace: &[Cycle],
+        bytecode_preprocessing: &BytecodePreprocessing,
+        sumcheck_challenges: &[F::Challenge],
+        gamma_powers: [F; 5],
+        r_cycle: &OpeningPoint<BIG_ENDIAN, F>,
+        r_product: &OpeningPoint<BIG_ENDIAN, F>,
+    ) -> Self {
+        let n_remaining_rounds = r_cycle.len() - sumcheck_challenges.len();
+        let r_prefix: OpeningPoint<BIG_ENDIAN, F> =
+            OpeningPoint::<LITTLE_ENDIAN, F>::new(sumcheck_challenges.to_vec()).match_endianness();
+
+        // Gen eq+1(r_cycle, (r_prefix, j)) for all j.
+        let EqPlusOnePrefixSuffixPoly {
+            prefix_0,
+            suffix_0,
+            prefix_1,
+            suffix_1,
+        } = EqPlusOnePrefixSuffixPoly::new(r_cycle);
+        let prefix_0_eval = MultilinearPolynomial::from(prefix_0).evaluate(&r_prefix.r);
+        let prefix_1_eval = MultilinearPolynomial::from(prefix_1).evaluate(&r_prefix.r);
+        let eq_plus_one_r_cycle: MultilinearPolynomial<F> = (0..suffix_0.len())
+            .map(|i| prefix_0_eval * suffix_0[i] + prefix_1_eval * suffix_1[i])
+            .collect::<Vec<F>>()
+            .into();
```

Comment:
You can evaluate `scaling_factor = Some(EqPlusOne::mle(r_cycle[..r_prefix.len()], r_prefix))`. And then just call `EqPlusOne::evals(r_cycle[r_prefix.len()..], scaling_factor)` to get `eq_plus_one_r_cycle`.